### PR TITLE
chore: modify dockerfile to use official tomcat images

### PR DESCRIPTION
**Pull request by varl**
This changes the Tomcat image we use in the second phase of the build to one of the officially supported Tomcat Docker images: https://hub.docker.com/_/tomcat

This makes it easier to just change the line:

```dockerfile
FROM tomcat:8.5.46-jdk8-openjdk-slim as serve
```

This allows us to test e.g. that DHIS2 works on Tomcat 9 quite easily:

```dockerfile
FROM tomcat:9.0.26-jdk8-openjdk-slim as serve
```

### Nota bene

⚠️ This does increase the image size to **587 MB** from **372 MB**.

Not sure what optimizations we can do though, I ran a couple of analysis tools on it to catch obvious stuff:

This Debian Slim image:

```
Total Image size: 587 MB
Potential wasted space: 6.8 MB
Image efficiency score: 99 %
```

Existing Alpine based image:

```
Total Image size: 372 MB
Potential wasted space: 1.4 MB
Image efficiency score: 99 %
```

ℹ️ _However_ it is important to note that the real value of this change is to be able to test the DHIS2 software against different versions of Tomcat.

### PR Updates

- [x] Fix so the permissions on the `temp/` dir matches the previous Alpine image
- [x] Add tomcat system-user to `at.deny` and `cron.deny`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -62,19 +62,17 @@ RUN rm -rf /usr/local/tomcat/webapps/* && \
     mkdir /usr/local/tomcat/webapps/ROOT && \
     chmod +rx /usr/local/bin/docker-entrypoint.sh && \
     mkdir $DHIS2_HOME && \
-    addgroup -S tomcat && \
-    addgroup root tomcat && \
```
Comment (varl) [inline Dockerfile]: The `root` user technically does not need to be in any group, but I'm not aware of why this was added in the first place so let me know why it's required and I'll add it back.

File: docker-entrypoint.sh [hunk 2]
```diff
@@ -18,7 +18,7 @@ if [ "$(id -u)" = "0" ]; then
         $TOMCATDIR/logs
 
     chown -R tomcat:tomcat $DHIS2HOME
-    exec su-exec tomcat "$0" "$@"
+    exec setpriv --reuid=tomcat --regid=tomcat --init-groups "$0" "$@"
```
Comment (varl) [inline docker-entrypoint.sh]: `setpriv` from `util-linux` offers the same functionality as `su-exec` but where `su-exec` is only in the main repo for alpine, `setpriv` is standard for Debian images.

Comment (amcgee) [reply docker-entrypoint.sh]: 👍

File: Dockerfile (hunk 1 above)
Comment (amcgee) [reply Dockerfile]: I don't know why it was added, but agree it probably isn't needed?  @vilkg ?

Comment (vilkg) [reply Dockerfile]: It was easier to manage access this way. If this change works the same way (container is started with non root user with the same read permissions on the same folders), I'm good

**Comment by vilkg** [comment]
@varl will test it today to see if security requirements haven't changed and tests passes!

**Comment by bobjolliffe** [comment]
Looking good.  Is there a reason that we open an AJP connector port in
server.xml?  I can see there might be, but I usually stick with one or the
other.

Does alpine include cron & at?  I think probably not, but no harm done to
put tomcat in /etc/at.deny and /etc/cron.deny in case it ever pops up
unexpected.  Those have been our common entry points for remote code
execution attacks in the past.

**Comment by varl** [comment]
> I think probably not, but no harm done to put tomcat in /etc/at.deny and /etc/cron.deny in case it ever pops up unexpected. Those have been our common entry points for remote code execution attacks in the past.

Thanks, updated. 👍 

> Looking good. Is there a reason that we open an AJP connector port in server.xml? I can see there might be, but I usually stick with one or the other.

Not sure, I'd rather leave the `server.xml` untouched for this PR and have a strict scope on only changing from the Alpine-based Tomcat image to the standard Debian-based Tomcat images.

We can take a pass at the `server.xml` in a different PR.

File: Dockerfile [hunk 3]
```diff
@@ -52,7 +52,7 @@ RUN mvn clean install -T1C -U -f /src/dhis-2/dhis-web/pom.xml -DskipTests
 # Serve the packaged .war file (Tomcat)
 # Use a second build stage in a fresh container so we don't bloat it with the Maven build tools
 ##########
-FROM tomcat:8.5.34-jre8-alpine as serve
+FROM tomcat:8.5.46-jdk8-openjdk-slim as serve
```
Comment (varl) [inline Dockerfile]: Perhaps we shouldn't do this in this PR but I'd like us to use the `tomcat:8.5-jdk-openjdk-slim` image to make sure that when they update Tomcat 8.5.x we get those fixes built-in without having to manually bump.

cc: @vilkg @Philip-Larsen-Donnelly

**Comment by varl** [comment]
@vilkg This is ready for testing.

**Comment by varl** [comment]
I was talking to @vilkg on Slack about naming conventions, specifically if we think that it's a problem for our end-users that we are switching our base image from Alpine to Debian, and bumping the Tomcat version from 8.5.34 to 8.5.46.

On one hand, we provide these images to abstract away the gritty details about what version of Tomcat we recommend people use, how to set up `server.xml`, and deploy it to a container. We take care of that so they don't have to.

Based on that there is an argument for just publishing new images using the same scheme we already have in place, since that means that by just updating the image they get the new best practices as well (security fixes, etc.) in a complete package.

On the other hand, if someone uses _our_ images as their base image, and has written customizations to it that are tailored for Alpine, their setup is going to break when we switch our base image to Debian.

The first argument is stronger than the second. If it becomes a problem for someone, it is going to be annoying for them to debug, so if we can mitigate that somehow it would be nice.

We have a decision on the [naming convention](https://github.com/dhis2/notes/blob/master/decisions/2019/05/25-war-docker-schemes.md#docker) itself, so that's not up for debate.

cc: @amcgee @Philip-Larsen-Donnelly

**Comment by vilkg** [comment]
I would like us to re-tag the existing images in core-dev channel to something like "master-tomcat-8.5.34-jre8-alpine", while leaving the default ones until it gets overwritten by the new debian based images. We could update the docs in docker hub to indicate, that the default images are Debian based. But I'm a bit unsure about the core channel: we won't rebuilt the images I suppose, so there's going to be mismatch between default images for new and old releases and apart of image size, it won't be noticeable for consumers right away. 

It might not be a big deal at all, but I would like us to be explicit and somewhat consistent about what we provide in the images.

**Comment by varl** [comment]
> I would like us to re-tag the existing images in core-dev channel to something like "master-tomcat-8.5.34-jre8-alpine"

> We could update the docs in docker hub to indicate, that the default images are Debian based.

Makes sense to me.

> But I'm a bit unsure about the core channel: we won't rebuilt the images I suppose, so there's going to be mismatch between default images for new and old releases and apart of image size, it won't be noticeable for consumers right away.

Couldn't we do it the same way as your suggestion for core-dev?

Add variant tags for the current images in addition to keeping the ones we have around, and then when we release the next version (e.g. 2.33.1) it will be built using a Debian based Tomcat?

So if a user pulls `dhis2/core:2.33.0` that would be the same as pulling `dhis2/core:2.33.0-tomcat-8.5.34-jre8-alpine`.

If a user then pulls `dhis2/core:2.33.1`, then that's the same as pulling `dhis2/core:2.33.1-tomcat-8.5.46-jdk8-slim` (if we also tag new "latest" releases as the exact variant).

>I would like us to be explicit and somewhat consistent about what we provide in the images.

👍  x 💯

**Comment by varl** [comment]
I have tagged the `dhis2/core-dev` images locally:

```
REPOSITORY          TAG                                IMAGE ID            CREATED             SIZE
dhis2/core-dev      2.32                               bc9311e5bfe2        2 hours ago         386MB
dhis2/core-dev      2.32-tomcat-8.5.34-jre8-alpine     bc9311e5bfe2        2 hours ago         386MB
dhis2/core-dev      2.31                               8de59a95aa26        24 hours ago        372MB
dhis2/core-dev      2.31-tomcat-8.5.34-jre8-alpine     8de59a95aa26        24 hours ago        372MB
dhis2/core-dev      master                             6b922513bf27        25 hours ago        372MB
dhis2/core-dev      master-tomcat-8.5.34-jre8-alpine   6b922513bf27        25 hours ago        372MB
dhis2/core-dev      2.33                               79d80ae9aae3        25 hours ago        372MB
dhis2/core-dev      2.33-tomcat-8.5.34-jre8-alpine     79d80ae9aae3        25 hours ago        372MB
dhis2/core-dev      2.30                               9953bcf6a19c        4 days ago          361MB
dhis2/core-dev      2.30-tomcat-8.5.34-jre8-alpine     9953bcf6a19c        4 days ago          361MB
```

Note the identical image id for each pair.

Doing the same for `dhis2/core` makes sense too:

```
REPOSITORY          TAG                                IMAGE ID            CREATED             SIZE
dhis2/core          2.33.0                             cb1cd9cebcb0        4 days ago          372MB
dhis2/core          2.33.0-tomcat-8.5.34-jre8-alpine   cb1cd9cebcb0        4 days ago          372MB
dhis2/core          2.32.2                             e15c45dc2daa        4 weeks ago         385MB
dhis2/core          2.32.2-tomcat-8.5.34-jre8-alpine   e15c45dc2daa        4 weeks ago         385MB
dhis2/core          2.31.5                             26d7f3548aff        8 weeks ago         370MB
dhis2/core          2.31.5-tomcat-8.5.34-jre8-alpine   26d7f3548aff        8 weeks ago         370MB
dhis2/core          2.32.1                             b6fbdb9347f6        2 months ago        385MB
dhis2/core          2.32.1-tomcat-8.5.34-jre8-alpine   b6fbdb9347f6        2 months ago        385MB
dhis2/core          2.31.4                             0fcd83bf2290        3 months ago        371MB
dhis2/core          2.31.4-tomcat-8.5.34-jre8-alpine   0fcd83bf2290        3 months ago        371MB
dhis2/core          2.32.0                             08127f0d6c15        4 months ago        381MB
dhis2/core          2.32.0-tomcat-8.5.34-jre8-alpine   08127f0d6c15        4 months ago        381MB
dhis2/core          2.31.3                             b1059961180c        4 months ago        364MB
dhis2/core          2.31.3-tomcat-8.5.34-jre8-alpine   b1059961180c        4 months ago        364MB
dhis2/core          2.31.2                             1dfd4acfc937        4 months ago        366MB
dhis2/core          2.31.2-tomcat-8.5.34-jre8-alpine   1dfd4acfc937        4 months ago        366MB
dhis2/core          2.31.1                             e926e7c2316b        4 months ago        366MB
dhis2/core          2.31.1-tomcat-8.5.34-jre8-alpine   e926e7c2316b        4 months ago        366MB
dhis2/core          2.31.0                             57493cd8eb79        4 months ago        364MB
dhis2/core          2.31.0-tomcat-8.5.34-jre8-alpine   57493cd8eb79        4 months ago        364MB
```

So the next release would have a default container at e.g. `:2.33.1` and the potential variants as well.

**Comment by vilkg** [comment]
@varl, yes, sounds good!

**Comment by varl** [comment]
Done. Merging.

**Pull request merged**